<a href="https://colab.research.google.com/github/JeonJeongumggumg/PyTorch/blob/main/predictTodata.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Training Data를 통해 입력데이터의 정답 출력

조건: 입력데이터가 12이하라면 정답이 0, 14이상이라면 정답 1을 출력

In [1]:
import torch

x_train = torch.Tensor([2,4,6,8,10,12,14,16,18,20]).view(10,1) #10행 1열로 reshape
y_train = torch.Tensor([0,0,0,0,0,0,1,1,1,1]).view(10,1)

print(x_train.shape, y_train.shape)

torch.Size([10, 1]) torch.Size([10, 1])


In [2]:
from torch import nn

class Model(nn.Module):
  def __init__(self):
    super().__init__()
    self.deeplearning_stack = nn.Sequential(
        nn.Linear(1,8), #1개의 입력데이터에 대해 8개의 출력을 내놓는 출력층
        nn.Linear(8,1), #8개의 입력데이터에 대해 1개의 출력을 내놓는 출력층
        nn.Sigmoid()  #sigmoid 활성화함수
    )

  def forward(self, data):
    prediction = self.deeplearning_stack(data)
    return prediction

In [3]:
deeplearning_model = Model() #모델 인스턴스 생성

for name, child in deeplearning_model.named_children(): #모델파라미터(가중치, 바이어스)값 확인
  for param in child.parameters():
    print(name, param)

deeplearning_stack Parameter containing:
tensor([[-0.0808],
        [ 0.4044],
        [-0.5509],
        [-0.8902],
        [-0.1141],
        [-0.9008],
        [-0.5470],
        [ 0.7171]], requires_grad=True)
deeplearning_stack Parameter containing:
tensor([-0.5083,  0.6244, -0.2802, -0.3081, -0.9051,  0.9749,  0.3326,  0.5202],
       requires_grad=True)
deeplearning_stack Parameter containing:
tensor([[ 0.2035, -0.1220, -0.2783, -0.1150,  0.3034, -0.2437, -0.1511, -0.2084]],
       requires_grad=True)
deeplearning_stack Parameter containing:
tensor([-0.2737], requires_grad=True)


입력층-은닉층 가충치 8개

은닉층 바이어스 값 8개

은닉층-출력층 가중치 8개

출력층 바이어스 1개

In [4]:
loss_function = nn.BCELoss() #손실함수는 이진분류이므로 Binary crossEntropy인 BCELoss()

optimizer = torch.optim.SGD(deeplearning_model.parameters(), lr=1e-1) #옵티마이저 SGD

In [6]:
nums_epoch = 6000

for epoch in range(nums_epoch+1):

  outputs = deeplearning_model(x_train) #입력데이터에 대해 예측값 계산

  loss = loss_function(outputs, y_train) #모델의 예측값과 정답과의 오차인 손실함수를 계산

  optimizer.zero_grad() #역전파코드, 학습이 진행됨에 따라 모델 파라미터를 업뎃하면서 최적화
  loss.backward()
  optimizer.step()

  if epoch % 100 == 0:
    print('epoch = ',epoch,'current loss = ', loss.item())

epoch =  0 current loss =  0.9200114011764526
epoch =  100 current loss =  0.3715370297431946
epoch =  200 current loss =  0.3014886975288391
epoch =  300 current loss =  0.2591528296470642
epoch =  400 current loss =  0.23002150654792786
epoch =  500 current loss =  0.2084580361843109
epoch =  600 current loss =  0.19170425832271576
epoch =  700 current loss =  0.1782277524471283
epoch =  800 current loss =  0.16709838807582855
epoch =  900 current loss =  0.1577156037092209
epoch =  1000 current loss =  0.14967183768749237
epoch =  1100 current loss =  0.14268004894256592
epoch =  1200 current loss =  0.13653060793876648
epoch =  1300 current loss =  0.1310696303844452
epoch =  1400 current loss =  0.12617790699005127
epoch =  1500 current loss =  0.12176255136728287
epoch =  1600 current loss =  0.11775077879428864
epoch =  1700 current loss =  0.1140848845243454
epoch =  1800 current loss =  0.1107175201177597
epoch =  1900 current loss =  0.10760913789272308
epoch =  2000 current 

In [7]:
deeplearning_model.eval() #신경망을 추론모드로 전환

test_data = torch.Tensor([0.5,3.0,3.5,11.0,13.0,31.0]).view(6,1)

pred = deeplearning_model(test_data) #6개의 testdata에 대해 예측을 실행

logical_value = (pred>0.5).float() #예측값이 0.5를 초과하면 true값을 실수 1.0으로 type casting/0.5이하라면 0.0으로 type casting

print(pred)
print(logical_value)

tensor([[2.5877e-16],
        [3.4473e-13],
        [1.4534e-12],
        [3.4244e-03],
        [5.2055e-01],
        [1.0000e+00]], grad_fn=<SigmoidBackward0>)
tensor([[0.],
        [0.],
        [0.],
        [0.],
        [1.],
        [1.]])


입력데이터가 12이하인 값에 대해선 0을 도출

13과 31의 예측값

13에 대해서는 약 52%확률로 1을 예측하고 있음